src/lang/std/engineConnection.ts

```diff
@@ -210,99 +268,76 @@ class EngineConnection {
     if (next.type === EngineConnectionStateType.Disconnecting) {
       const sub = next.value
       if (sub.type === DisconnectingType.Error) {
+        console.log(sub)
+
         // Record the last step we failed at.
         // (Check the current state that we're about to override that
         // it was a Connecting state.)
-        console.log(sub)
         if (this._state.type === EngineConnectionStateType.Connecting) {
-          if (!sub.value) sub.value = {}
+          if (!sub.value) sub.value = { error: ConnectionError.Unknown }
           sub.value.lastConnectingValue = this._state.value
         }
 
         console.error(sub.value)
       }
     }
     this._state = next
-    this.onConnectionStateChange(this._state)
+
+    this.dispatchEvent(
+      new CustomEvent(EngineConnectionEvents.ConnectionStateChanged, {
+        detail: this._state,
+      })
+    )
```

Comment:
I don't love using such a browser-specific feature, as eventually, we'll want to migrate as much of this as possible rust/wasm. That was the motivation for not integrating this into Xstate/the rest of the app's state management and using callbacks because class and callbacks are pretty generic and will much more easily convert to rust.

I get this is a vague hypothetical without a date on when we'll carry out this work 🤷 

What was the motivation for using Events again? (I say again because @paultag gravitated towards them too).

Reply:
The motivation for using events is we were already using events for this before, but instead this uses the Events API instead of a new events API implementation. We already had instances of the code creating subscribers.

Additionally, it makes sense to use events as a programming method here because many of the objects we communicate also use this event-based paradigm. Having a paradigm change creates a weird interface we have to maintain and further think about how everything maps or connects together.

The Events API is not browser specific: https://developer.mozilla.org/en-US/docs/Web/API/Event#browser_compatibility you can see Deno and NodeJS support this API.

Rust has many event based packages, I'm not sure why it would matter if we used a callback based vs event based API:

* https://crates.io/crates/presage
* https://crates.io/crates/eventify
* https://crates.io/crates/event
* https://crates.io/crates/statig

Additionally even if we want move engineConnection into Rust, it may not be the best idea, because there are many browser related mechanisms we use already.

We may already be past the point of "we want it all in Rust". It's very possible the future of this looks like a complete rewrite.

I hope that clarifies why I've gone with the Events API.

Reply:
As I'm sitting here at the airport I think I've thought of further "why".

Callbacks are usually for a specific call-site and only one can be bound.

Events can be subscribed to from anywhere and many can be bound.

We can probably "downgrade" to callbacks (I don't mean this as a bad thing) with some careful thinking to make this _perfect_.

There is some level of simultaneous operations that are going on in engineConnection, but I also understand we actually care about mostly sequencing it instead. This would be better to address during a rewrite to Rust.

I would further say that writing JS/TS in an idiomatic way is much easier to maintain than in a Rust idiomatic way _in JS/TS_ because there will be concepts which are weird to express in one or the other. (Just look at our C++-Rust bridge which are arguably closer to eachother  in capabilities...)